camkes/parser/fdtQueryEngine.py

```diff
@@ -414,12 +415,137 @@ def __init__(self):
+        result.append(x & ((1 << 32) - 1))
+        return result
+
+    @staticmethod
+    def xlat_reg(addr, size, buses):
+        for bus in buses:
+            if not 'translations' in bus:
+                continue
+            for xlat in bus['translations']:
+                if addr < xlat['from_start'] or addr > xlat['from_start'] + xlat['size'] - 1:
+                    continue
+                if addr + size > xlat['from_start'] + xlat['size']:
+                    raise RuntimeError('Region 0x%x/0x%x does not fit into range 0x%x/0x%x'
+                        % (addr, size, xlat['from_start'], xlat['size']))
+                    continue
+                addr = xlat['to_start'] + (addr - xlat['from_start'])
+                break
+        return addr
+
+    @classmethod
+    def xlat_regs(cls, cells):
+        if not 'reg' in cells:
+            return
+
+        this_address_cells = cells['buses'][0]['this-address-cells']
+        this_size_cells = cells['buses'][0]['this-size-cells']
+
+        regs = cls.unpacker(cells['reg'], [this_address_cells, this_size_cells])
+        cells['reg'] = []
+
+        while True:
+            (addr, size) = next(regs, (None, None))
+            if addr == None:
+                break
+
+            addr = cls.xlat_reg(addr, size, cells['buses'])
+
+            cells['reg'].extend(cls.pack(addr, this_address_cells))
+            cells['reg'].extend(cls.pack(size, this_size_cells))
+
+    @classmethod
+    def resolve_translations(cls, buses):
+        # find out necessary address range translations
+        bus_iter = iter(buses)
+        this_bus = next(bus_iter, None)
+
+        while this_bus:
+            parent_bus = next(bus_iter, None)
+            if 'ranges' in this_bus:
+                if not parent_bus:
+                    raise RuntimeError('cannot have \'ranges\' without parent bus')
+
+                # deserialize the 32-bit words from device tree
+                ranges = cls.unpacker(this_bus['ranges'], [
+                    this_bus['this-address-cells'],
+                    parent_bus['this-address-cells'],
+                    this_bus['this-size-cells'],
+                ])
+
+                this_bus['translations'] = []
+                while True:
+                    (this_addr, parent_addr, this_size) = next(ranges, (None, None, None))
+                    if this_addr == None:
+                        break
+                    this_bus['translations'].append({
+                        'from_start': this_addr,
+                        'to_start': parent_addr,
+                        'size': this_size,
+                    })
+
+            this_bus = parent_bus
+
+        return buses
 
-        address_cells_key = 'this-address-cells'
-        size_cells_key = 'this-size-cells'
-        node_path_key = 'this-node-path'
+    @staticmethod
+    def resolve_buses(node):
+        buses = []
+        parent = node.parent
+        while True:
```

Comment:
This loop is reorganized so that a node is guaranteed to have at least one bus. And in case it does not explicitly tell the sizes of address and size cells, default them to values as per in spec. So in other words, resolved['buses'][0] is now always guaranteed to exist.